### PR TITLE
(NETDEV-30) Enhance syslog, tacacs, radius

### PR DESCRIPTION
**Pull request by shermdog**
This PR enhances the existing syslog, tacacs, and radius types as specified in  puppetlabs/netdev_stdlib#25

syslog_server
- port

syslog_settings
- console
- monitor
- source_interface
- vrf

tacacs_global
- source_interface

radius_global
- source_interface

syslog_server initialize now uses options hash
  * Prior to this release ntp_server accepted positional arguments for id and
  prefer.  New behavior is to pass attributes as a hash.

  Example:
  ```
 options = { 'name' => '1.1.1.1', 'level' => '4', 'port' => '2154',
             'vrf' => 'red' }
 Cisco::SyslogServer.new(options, true)
  ```

* tacacs_global key removal fixed
  * Prior to this release key removal was done by passing in a value of 8.  A
  nil value is now used.  Added intelligence to determine key format
  automatically for removal.

### CODE REVIEW

File: cisco_node_utils.gemspec [hunk 1]
```diff
@@ -31,6 +31,7 @@ Currently supports NX-OS nodes.
   spec.add_development_dependency 'bundler', '~> 1.7'
   spec.add_development_dependency 'kwalify', '~> 0.7.2'
   spec.add_development_dependency 'minitest', '~> 5.0'
+  spec.add_development_dependency 'net-telnet', '~> 0.1.1'
```
Comment (mikewiebe) [inline cisco_node_utils.gemspec:34]: Why is this needed?

File: lib/cisco_node_utils/radius_global.rb [hunk 2]
```diff
@@ -137,5 +137,29 @@ def key_set(value, format)
...
+      else
+        config_set(
+          'radius_global', 'source_interface',
+          state: 'no', source_interface: '')
+      end
+    end
+
+    # Get source interface
+    def source_interface
+      config_get('radius_global', 'source_interface') ||
```
Comment (mikewiebe) [inline lib/cisco_node_utils/radius_global.rb]: Any time we get interface information we downcase it to normalize the data.
```ruby
config_get('radius_global', 'source_interface').downcase
```

File: lib/cisco_node_utils/radius_global.rb [hunk 3]
```diff
@@ -137,5 +137,29 @@ def key_set(value, format)
...
+    end
+
+    # Set source interface
+    def source_interface=(name)
+      if name
+        config_set(
+          'radius_global', 'source_interface',
+          state: '', source_interface: name)
+      else
+        config_set(
```
Comment (mikewiebe) [inline lib/cisco_node_utils/radius_global.rb:153]: Is your intent to use `default` as the manifest setting to remove the source_interface?

File: lib/cisco_node_utils/syslog_settings.rb [hunk 4]
```diff
@@ -42,6 +42,68 @@ def ==(other)
       name == other.name
     end
 
+    def console
```
Comment (mikewiebe) [inline lib/cisco_node_utils/syslog_settings.rb]: You should add default getters here for `console, monitor and source_interface` like you did for radius_global.

File: lib/cisco_node_utils/syslog_settings.rb [hunk 5]
```diff
@@ -42,6 +42,68 @@ def ==(other)
...
+          state: '', severity: severity)
+      else
+        config_set(
+          'syslog_settings', 'monitor',
+          state: 'no', severity: '')
+      end
+    end
+
+    def source_interface
+      config_get('syslog_settings', 'source_interface')
```
Comment (mikewiebe) [inline lib/cisco_node_utils/syslog_settings.rb]: config_get('syslog_settings', 'source_interface').downcase

File: lib/cisco_node_utils/syslog_settings.rb [hunk 6]
```diff
@@ -42,6 +42,68 @@ def ==(other)
...
+      if console.is_a?(Array)
+        console = console[0] == 'no' ? 'unset' : console[1]
+      end
+      console
+    end
+
+    def console=(severity)
+      if severity
+        fail ArgumentError,
+             'Console severity must be in range [0-7]' \
```
Comment (mikewiebe) [inline lib/cisco_node_utils/syslog_settings.rb]: In general for ranges like this we just let the cli reject it.  That way if the range changes in the future we don't have to update the object.  The exception of course is if the message that gets returned is not at all meaningful.  Here it's quasi meaningful :).

```shell
dt-n9k5-1(config)# logging console ?
  <CR>   
  <0-7>  0-emerg;1-alert;2-crit;3-err;4-warn;5-notif;6-inform;7-debug

dt-n9k5-1(config)# logging console 9
                                   ^
% Invalid number, range is (0:7) at '^' marker.
dt-n9k5-1(config)# 

```

File: lib/cisco_node_utils/syslog_settings.rb [hunk 7]
```diff
@@ -42,6 +42,68 @@ def ==(other)
...
+      if monitor.is_a?(Array)
+        monitor = monitor[0] == 'no' ? 'unset' : monitor[1]
+      end
+      monitor
+    end
+
+    def monitor=(severity)
+      if severity
+        fail ArgumentError,
+             'Monitor severity must be in range [0-7]' \
```
Comment (mikewiebe) [inline lib/cisco_node_utils/syslog_settings.rb]: Same comment as above about range checking.

File: lib/cisco_node_utils/tacacs_global.rb [hunk 8]
```diff
@@ -92,14 +88,47 @@ def self.default_key
...
+      else
+        config_set(
+          'tacacs_global', 'source_interface',
+          state: 'no', source_interface: '')
       end
     end
+
+    # Get source interface
+    def source_interface
+      config_get('tacacs_global', 'source_interface') ||
```
Comment (mikewiebe) [inline lib/cisco_node_utils/tacacs_global.rb]: config_get('tacacs_global', 'source_interface').downcase

File: tests/test_syslog_settings.rb [hunk 9]
```diff
@@ -64,4 +67,38 @@ def test_create
                   )
     end
   end
+
+  def test_console
+    syslog_setting = Cisco::SyslogSettings.new('default')
+
+    assert_equal(2, syslog_setting.console)
+    syslog_setting.console = '5'
```
Comment (mikewiebe) [inline tests/test_syslog_settings.rb]: better to use `syslog_setting.console = '1'`.  You will get the following error if you use a value of `5` and the baud rate is not compatible.

```shell
Node under test:
  - name  - dt-n9k5-1
  - type  - N9K-C9396PX
  - image - bootflash:///nxos.7.0.3.I6.1.bin

TestSyslogSettings#test_source_interface = 14.70 s = .
TestSyslogSettings#test_timestamp = 9.58 s = .
TestSyslogSettings#test_monitor = 10.89 s = .
TestSyslogSettings#test_console = 9.68 s = E

Finished in 44.858746s, 0.0892 runs/s, 0.3121 assertions/s.

  1) Error:
TestSyslogSettings#test_console:
Cisco::CliError: [#<Cisco::SyslogSettings:0x007f0d6b1a6038>] The command ' logging console 5' was rejected with error:
Baud rate of console should be 38400 to increase severity level

    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/client/nxapi/client.rb:298:in `handle_output'
    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/client/nxapi/client.rb:229:in `req'
    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/client/nxapi/client.rb:130:in `set'
    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/node.rb:276:in `set'
    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/node.rb:216:in `config_set'
    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/node_util.rb:72:in `config_set'
    /nobackup/mwiebe/NU/shermdog/cisco-network-node-utils/lib/cisco_node_utils/syslog_settings.rb:58:in `console='
    tests/test_syslog_settings.rb:75:in `test_console'

4 runs, 14 assertions, 0 failures, 1 errors, 0 skips
```

File: cisco_node_utils.gemspec (hunk 1 above)
Comment (shermdog) [reply cisco_node_utils.gemspec:34]: It's required by basetest

https://github.com/cisco/cisco-network-node-utils/blob/develop/tests/basetest.rb#L32

Comment (mikewiebe) [reply cisco_node_utils.gemspec:34]: Good catch.  Thanks for adding that in.

File: lib/cisco_node_utils/syslog_settings.rb (hunk 6 above)
Comment (shermdog) [reply lib/cisco_node_utils/syslog_settings.rb]: I can take this out, but the levels are defined by RFC 5424, so I think the likelihood of them changing in the next 5 years is low.  We're enforcing the value input on the puppet side, so it's entirely your call.

File: lib/cisco_node_utils/radius_global.rb (hunk 3 above)
Comment (shermdog) [reply lib/cisco_node_utils/radius_global.rb:153]: On the puppet side we're looking for `'unset'`, which would then pass `nil` to node_utils.  For whatever reason that's what the netdev providers/utils have been doing.

Comment (mikewiebe) [reply lib/cisco_node_utils/radius_global.rb:153]: Interesting.  Currently the types (https://github.com/puppetlabs/netdev_stdlib/pull/25/files) only indicate that a String is accepted.  I think it would be clearer if we updated the input validation to indicate a string representing the interface or the string 1`'unset'` to remove the interface.  Thoughts?

File: lib/cisco_node_utils/syslog_settings.rb (hunk 6 above)
Comment (mikewiebe) [reply lib/cisco_node_utils/syslog_settings.rb]: If the enforcement is happening on the puppet side then I certainly don't think we need to do that here.  Lets keep it simple on this side.

File: lib/cisco_node_utils/radius_global.rb (hunk 3 above)
Comment (shermdog) [reply lib/cisco_node_utils/radius_global.rb:153]: I'll have to see how things are being implemented with other vendor modules.

**Comment by mikewiebe** [comment]
👍   Will merge **AFTER** puppetlabs/netdev_stdlib#25 is merged.

**Comment by shermdog** [comment]
@mikewiebe netdev_stdlib PR has been merged, and module has been released as 0.13.0
